### PR TITLE
Ind. LabelModel Part 2: Add intermediate tests for label model

### PR DESCRIPTION
**Pull request by paroma**
Stacked PR part 2 (Part 1: #1209) stemming from #1182 

* Add high coverage, intermediate function tests for label model

* Only include conditionally independent testing

### CODE REVIEW

File: snorkel/labeling/model/label_model.py [hunk 1]
```diff
@@ -121,14 +121,14 @@ def _build_mask(self):
...
-    def _generate_O(self, L):
+    def _generate_O(self, L, higher_order=False):
         """Form the overlaps matrix, which is just all the different observed
         combinations of values of pairs of sources
 
         Note that we only include the k non-abstain values of each source,
         otherwise the model not minimal --> leads to singular matrix
         """
-        L_aug = self._get_augmented_label_matrix(L)
+        L_aug = self._get_augmented_label_matrix(L, higher_order=higher_order)
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: why is this change happening here? is it just to facilitate testing?

File: test/labeling/model/test_label_model.py [hunk 2]
```diff
@@ -79,6 +115,115 @@ def test_augmented_L_construction(self):
...
+        L = np.array([[1, 0, 1], [1, 2, 1]])
+        label_model = self._set_up_model(L)
+
+        label_model.train_model(L, n_epochs=1)
+        init_loss = label_model.loss_mu().detach().numpy().ravel()[0]
+
+        label_model.train_model(L, n_epochs=10)
+        next_loss = label_model.loss_mu().detach().numpy().ravel()[0]
+
+        self.assertLessEqual(next_loss, init_loss)
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: Repeating myself, but isn't this highly dependent on default hyperparams? Should we hardcode those (even just LR) into the call here in case they change?

File: test/labeling/model/test_label_model.py [hunk 3]
```diff
@@ -79,6 +115,115 @@ def test_augmented_L_construction(self):
...
+        np.testing.assert_array_almost_equal(probs, true_probs)
+
+    def test_loss(self):
+        L = np.array([[1, 0, 1], [1, 2, 0]])
+        label_model = self._set_up_model(L)
+        label_model._get_augmented_label_matrix(L, higher_order=True)
+
+        label_model.mu = nn.Parameter(label_model.mu_init.clone() + 0.05)
+        self.assertAlmostEqual(
+            label_model.loss_l2(l2=1.0).detach().numpy().ravel()[0], 0.03
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: How do we know these are the right values?

Comment (henryre) [reply test/labeling/model/test_label_model.py]: Comment about that would be good

File: test/labeling/model/test_label_model.py [hunk 4]
```diff
@@ -79,6 +115,115 @@ def test_augmented_L_construction(self):
...
+        label_model.mu = nn.Parameter(label_model.mu_init.clone())
+        accs = label_model.get_accuracies(probs=None)
+        np.testing.assert_array_almost_equal(accs, np.array([0.5, 0.355]))
+
+    def test_build_mask(self):
+
+        L = np.array([[1, 2, 1], [1, 2, 1]])
+        label_model = self._set_up_model(L)
+
+        true_mask = np.array(
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: Great!

File: test/labeling/model/test_label_model.py [hunk 5]
```diff
@@ -79,6 +115,115 @@ def test_augmented_L_construction(self):
...
+
+        label_model.m = 2
+        label_model.k = 2
+        label_model.P = torch.Tensor([[0.5, 0.0], [0.0, 0.5]])
+        accs = label_model.get_accuracies(probs=probs)
+        np.testing.assert_array_almost_equal(accs, np.array([0.7, 0.825]))
+
+        label_model.mu = nn.Parameter(label_model.mu_init.clone())
+        accs = label_model.get_accuracies(probs=None)
+        np.testing.assert_array_almost_equal(accs, np.array([0.5, 0.355]))
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: How do we know this is right?

File: test/labeling/model/test_label_model.py [hunk 6]
```diff
@@ -1,48 +1,84 @@
...
         )
+        np.testing.assert_array_almost_equal(label_model.O.numpy(), true_O)
+
+        L = np.array([[1, 2, 1], [1, 2, 1], [2, 1, 1], [1, 2, 2]])
+        label_model = self._set_up_model(L)
+
+        label_model._generate_O(L, higher_order=False)
+        true_O = np.array(
+            [
+                [3 / 4, 0, 0, 3 / 4, 1.0 / 2, 1 / 4],
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: nit: 1.0

File: test/labeling/model/test_label_model.py [hunk 7]
```diff
@@ -1,48 +1,84 @@
...
+        label_model._set_class_balance(class_balance=None, Y_dev=Y_dev)
+        np.testing.assert_array_almost_equal(label_model.p, np.array([0.6, 0.4]))
+
+    def test_generate_O(self):
+        L = np.array([[1, 2, 1], [1, 2, 1], [2, 1, 1], [1, 2, 2]])
+        label_model = self._set_up_model(L)
+
+        true_O = np.array(
+            [
+                [3 / 4, 0, 0, 3 / 4, 1.0 / 2, 1 / 4],
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: nit: 1.0

File: snorkel/labeling/model/label_model.py (hunk 1 above)
Comment (paroma) [reply snorkel/labeling/model/label_model.py]: yes, facilitating testing + for later, we might make the O_inv method public and want to pass this argument in

File: test/labeling/model/test_label_model.py (hunk 2 above)
Comment (paroma) [reply test/labeling/model/test_label_model.py]: my bad, ill hardcode the params to be safe

File: test/labeling/model/test_label_model.py (hunk 3 above)
Comment (paroma) [reply test/labeling/model/test_label_model.py]: Wolfram alpha! will add math comments

Comment (ajratner) [reply test/labeling/model/test_label_model.py]: Definitely add comments- esp in tests as people often read to understand what's going on!
